src/utilities/substituteFragmentArguments.ts

```diff
@@ -0,0 +1,78 @@
+import type { Maybe } from '../jsutils/Maybe.js';
+import type { ObjMap } from '../jsutils/ObjMap.js';
+
+import type {
+  ArgumentNode,
+  FragmentDefinitionNode,
+  FragmentSpreadNode,
+  SelectionSetNode,
+  ValueNode,
+  VariableDefinitionNode,
+} from '../language/ast.js';
+import { Kind } from '../language/kinds.js';
+import { visit } from '../language/visitor.js';
+
+/**
+ * Replaces all fragment argument values with non-fragment-scoped values.
+ *
+ * NOTE: fragment arguments are scoped to the fragment they're defined on.
+ * Therefore, after we apply the passed-in arguments, all remaining variables
+ * must be either operation defined variables or explicitly unset.
+ */
+export function substituteFragmentArguments(
+  def: FragmentDefinitionNode,
+  fragmentSpread: FragmentSpreadNode,
+): SelectionSetNode {
+  const argumentDefinitions = def.variableDefinitions;
```

Comment:
Agree it's not super clean. I likely need to do another pass but there is some logic here: `argumentDefinitions` are the definitions for the arguments we need to look up based on the spread, *not* the variable definitions that the spread currently has access to.

Reply:
I don't know that I follow that logic exactly for this specific use. In our existing nomenclature, arguments are the input values you provide (to a field, directive, now frag spread) and variables are the placeholders a value will go to.

I do get the logic that "variable definition" does some double-duty here. It defines the variables within fragment but it also defines the arguments allowed to that fragment's spreads. For consistency we should choose one naming convention and use comments when that could cause confusion. Since the AST node is "variable definition" - I think we should stick with that.

In this particular case you're actually substituting/replacing the variables within a fragment with values which were provided via fragment spread arguments. We're not substituting/replacing any fragment spread arguments themselves.

The first step here `fragmentArgumentSubstitutions` seems to me to be logically taking input of "variable definitions" and "fragment spread arguments" and returning output of "variable values"


---

Tl;dr - naming things is hard